configs/configload/loader_load.go

```diff
@@ -0,0 +1,97 @@
+package configload
+
+import (
+	"fmt"
+
+	version "github.com/hashicorp/go-version"
+	"github.com/hashicorp/hcl2/hcl"
+	"github.com/hashicorp/terraform/configs"
+)
+
+// LoadConfig reads the Terraform module in the given directory and uses it as the
+// root module to build the static module tree that represents a configuration,
+// assuming that all required descendent modules have already been installed.
+//
+// If error diagnostics are returned, the returned configuration may be either
+// nil or incomplete. In the latter case, cautious static analysis is possible
+// in spite of the errors.
+//
+// LoadConfig performs the basic syntax and uniqueness validations that are
+// required to process the individual modules, and also detects
+func (l *Loader) LoadConfig(rootDir string) (*configs.Config, hcl.Diagnostics) {
```

Comment:
I'm not a fan of checking for a nil return value to indicate an error. Though it's documented here, I think this invites missed nil checks as the code evolves (which we've had a _lot_ of throughout other terraform code).

Can we make this so we use can rely on the diagnostics to convey the error condition?

Reply:
There is no case here where `nil` is returned without an error. The distinction between `nil` and an object here is an optional distinction that allows the caller to recognize a specific error situation, but if the caller doesn't care to make that distinction the usual pattern will work as normal:

```go
config, diags := l.LoadConfig(dir)
if diags.HasErrors() {
  // abort processing, and never look at `config` at all
}
```

The reason this is called out here is that in general HCL2 tries to follow the pattern that the returned object should always be populated as much as possible even if there are errors, so that the caller can inspect the possibly-partial object (carefully!) in that case. This function does not meet that expectation, so the documentation is calling that out to say that the caller _cannot_ expect that the returned object will be valid-but-incomplete in call cases, and so _unusual_ callers that _do_ want to inspect the erroneous result (e.g. for text-editor-integration use-cases like "Go to Definition", which should ideally work even when config is partially invalid) are aware that they need to accommodate a `nil` here.

```go
// **Unusual** usage for editor integrations
config, _ := l.LoadConfig(dir) // ignore diagnostics altogether, so must be more cautious
if config == nil {
  return (something representing "not found")
}
variable := config.Module.Variables[varName]
if variable == nil {
  return (something representing "not found")
}
return variable.DeclRange
```

Reply:
Sounds good, as long as we can maintain that `if diags.HashErrors()` is false, then the config is not nil. I'd like to at least start working against the "might be nil" returns, since they've caused quite a few bugs.

Reply:
Definitely... the general "rule" for functions that return `hcl.Diagnostics` is that the return value must be valid (whatever that means for the given function) _unless_ `diags.HasErrors`. If that isn't true anywhere then that's a bug to be fixed.